File: server/src/main/java/cn/iinti/atom/system/PrivacyFilter.java
package cn.iinti.atom.system;

import cn.iinti.atom.BuildConfig;
import cn.iinti.atom.service.base.config.Settings;
import jakarta.servlet.*;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import lombok.extern.slf4j.Slf4j;
import org.springframework.boot.web.servlet.FilterRegistrationBean;
import org.springframework.context.annotation.Bean;
import org.springframework.stereotype.Component;

import java.io.IOException;

@Slf4j
@Component
public class PrivacyFilter implements Filter {
    private static boolean isSwaggerAccess(String uri) {
        return uri.startsWith("/v3/api-docs")
                || uri.startsWith("/swagger-ui/");
    }

    private static boolean isActuator(String uri) {
        return uri.startsWith(BuildConfig.restfulApiPrefix + "/actuator");
    }

    @Override
    public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain) throws IOException, ServletException {
        HttpServletRequest httpServletRequest = (HttpServletRequest) request;
        String requestURI = httpServletRequest.getRequestURI();
        boolean block = Settings.blockSwagger.value && isSwaggerAccess(requestURI);
        if (!block && Settings.blockActuator.value && isActuator(requestURI)) {
            block = true;
        }
        if (!block) {
            chain.doFilter(request, response);
            return;
        }
        HttpServletResponse httpServletResponse = (HttpServletResponse) response;
        httpServletResponse.setStatus(404);
    }

    @Bean
    public FilterRegistrationBean<PrivacyFilter> privacyFilterRegistrationBean() {
        FilterRegistrationBean<PrivacyFilter> registration = new FilterRegistrationBean<>();
        registration.setFilter(this);
        registration.setName("privacyFilter");
        registration.addUrlPatterns("/*");
        registration.setOrder(1);
        return registration;
    }

}

File: server/src/main/kotlin/cn/iinti/atom/system/PrivacyFilter.kt
package cn.iinti.atom.system

import cn.iinti.atom.BuildInfo
import cn.iinti.atom.service.base.config.Settings
import jakarta.servlet.Filter
import jakarta.servlet.FilterChain
import jakarta.servlet.ServletRequest
import jakarta.servlet.ServletResponse
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.boot.web.servlet.FilterRegistrationBean
import org.springframework.context.annotation.Bean
import org.springframework.stereotype.Component

@Component
class PrivacyFilter : Filter {

    override fun doFilter(request: ServletRequest, response: ServletResponse, chain: FilterChain) {
        try {
            val httpServletRequest = request as HttpServletRequest
            val requestURI = httpServletRequest.requestURI
            var block = Settings.blockSwagger.value!! && isSwaggerAccess(requestURI)
            if (!block && Settings.blockActuator.value!! && isActuator(requestURI)) {
                block = true
            }
            if (!block) {
                chain.doFilter(request, response)
                return
            }
            val httpServletResponse = response as HttpServletResponse
            httpServletResponse.status = 404
        } catch (e: Exception) {
            // Ignore exception
        }
    }

    private fun isSwaggerAccess(uri: String): Boolean {
        return uri.startsWith("/v3/api-docs") || uri.startsWith("/swagger-ui/")
    }

    private fun isActuator(uri: String): Boolean {
        return uri.startsWith(BuildInfo.restfulApiPrefix + "/actuator")
    }

    @Bean
    fun privacyFilterRegistrationBean(): FilterRegistrationBean<PrivacyFilter> {
        val registration = FilterRegistrationBean<PrivacyFilter>()
        registration.filter = this
        registration.setName("privacyFilter")
        registration.addUrlPatterns("/*")
        registration.order = 1
        return registration
    }
}